In [5]:
import gymnasium as gym
from stable_baselines3 import PPO
from config import config
from stable_baselines3.common.utils import set_random_seed
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.evaluation import evaluate_policy
import optuna
from stable_baselines3.common.callbacks import EvalCallback,BaseCallback
from config2 import config2
env_id = "intersection-v0"#"racetrack-v0"
num_cpu = 4

# env = make_vec_env(env_id, n_envs=num_cpu)
# env = gym.make(env_id, render_mode="rgb_array", config=config
            #    )
# env.unwrapped.configure(config)

In [6]:
config

{'observation': {'type': 'GrayscaleObservation',
  'observation_shape': (64, 64),
  'stack_size': 4,
  'weights': [0.2989, 0.587, 0.114],
  'scaling': 1.75},
 'action': {'type': 'ContinuousAction', 'longitudinal': True, 'lateral': True},
 'duration': 13,
 'destination': 'o1',
 'initial_vehicle_count': 150,
 'spawn_probability': 0.7,
 'screen_width': 600,
 'screen_height': 600,
 'centering_position': [0.5, 0.5],
 'scaling': 7.15,
 'collision_reward': -5,
 'high_speed_reward': 0,
 'arrived_reward': 3,
 'normalize_reward': False,
 'simulation_frequency': 15,
 'policy_frequency': 5,
 'offroad_terminal': False}

In [7]:
model = PPO.load("models/optuna_last_chance_130000.zip")
eval_env = gym.make(env_id, render_mode="rgb_array", config=config)
# eval_env = gym.make("highway-fast-v0", render_mode="rgb_array", config=config)

In [8]:
while True:
  done = truncated = False
  obs, info = eval_env.reset()
  while not (done or truncated):
    action, _states = model.predict(obs, deterministic=True)
    obs, reward, done, truncated, info = eval_env.step(action)
    eval_env.render()

AttributeError: 'NoneType' object has no attribute 'get_image'

In [ ]:
class DoneCallback(BaseCallback):
    def __init__(self, check_freq: int, save_freq: int, save_path: str, verbose=1):
        super(DoneCallback, self).__init__(verbose)
        self.check_freq = check_freq
        self.save_freq = save_freq
        self.save_path = save_path
        self.done_count = 0

    def _on_step(self) -> bool:
        if 'done' in self.locals.keys():
            if self.locals['done']:
                self.done_count += 1

        if self.n_calls % self.check_freq == 0:
            print(f"Step: {self.num_timesteps} Done count: {self.done_count}")

        if self.n_calls % self.save_freq == 0:
            self.model.save(self.save_path + str(self.num_timesteps))

        return True

In [3]:
def optimize_ppo(trial):
    """ Learning hyperparameters we want to optimise"""
    return {
        'n_steps': int(trial.suggest_loguniform('n_steps', 32, 2048)),
        'gamma': trial.suggest_categorical('gamma', [0.9, 0.95, 0.98, 0.99, 0.999, 0.9999]),
        'learning_rate': trial.suggest_loguniform('learning_rate', 1e-5, 1e-2),
        'ent_coef': trial.suggest_loguniform('ent_coef', 0.00000001, 0.1),
        'clip_range': trial.suggest_uniform('clip_range', 0.1, 0.4),
        'n_epochs': int(trial.suggest_loguniform('n_epochs', 1, 10)),
    }

def objective(trial):
    """ Objective function for optimization """
    env = make_vec_env(env_id, n_envs=num_cpu,env_kwargs={"config":config})
    log_dir = "logs"
    model = PPO('CnnPolicy', env, verbose=0,tensorboard_log=log_dir,**optimize_ppo(trial))

    callback = DoneCallback(check_freq=128, save_freq=5000, save_path="./models/hypertuning_")

    # Create an evaluation environment
    eval_env = make_vec_env(env_id, n_envs=1,env_kwargs={"config":config})

    # Create the evaluation callback
    eval_callback = EvalCallback(eval_env, best_model_save_path='./logs/',
                                 log_path='./logs/', eval_freq=500,
                                 deterministic=True, render=False)

    # Train the model, passing both the training and evaluation callbacks
    model.learn(total_timesteps=512*50, callback=[callback, eval_callback], progress_bar=True)

    # Retrieve the best reward
    best_reward = eval_callback.best_mean_reward
    return best_reward

study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=30)

print("Number of finished trials: ", len(study.trials))
print("Best trial:")
trial = study.best_trial

print("  Value: ", trial.value)
print("  Params: ")
for key, value in trial.params.items():
    print("    {}: {}".format(key, value))

Eval num_timesteps=26000, episode_reward=18.20 +/- 4.66

Episode length: 30.00 +/- 18.26

[I 2024-04-25 13:44:12,889] Trial 25 finished with value: 27.0 and parameters: {'n_steps': 1087.4953333804424, 'gamma': 0.98, 'learning_rate': 6.598882302687722e-05, 'ent_coef': 0.01786356354334364, 'clip_range': 0.16873577386904987, 'n_epochs': 1.3576674975524343}. Best is trial 17 with value: 42.6.
c:\Users\matth\AppData\Local\Programs\Python\Python311\Lib\site-packages\stable_baselines3\ppo\ppo.py:155: UserWarning: You have specified a mini-batch size of 64, but because the `RolloutBuffer` is of size `n_steps * n_envs = 1072`, after every 16 untruncated mini-batches, there will be a truncated mini-batch of size 48
We recommend using a `batch_size` that is a factor of `n_steps * n_envs`.
Info: (n_steps=268 and n_envs=4)
  warnings.warn(


Output()

Step: 512 Done count: 3

Step: 1024 Done count: 6

Step: 1536 Done count: 10

Eval num_timesteps=2000, episode_reward=24.20 +/- 12.40

Episode length: 27.20 +/- 10.03

New best mean reward!

Step: 2048 Done count: 15

Step: 2560 Done count: 18

Step: 3072 Done count: 23

Step: 3584 Done count: 27

Eval num_timesteps=4000, episode_reward=17.60 +/- 4.84

Episode length: 38.60 +/- 22.46

Step: 4096 Done count: 32

Step: 4608 Done count: 37

Step: 5120 Done count: 42

Step: 5632 Done count: 44

Eval num_timesteps=6000, episode_reward=10.60 +/- 3.93

Episode length: 27.00 +/- 8.00

Step: 6144 Done count: 48

Step: 6656 Done count: 54

Step: 7168 Done count: 57

Step: 7680 Done count: 61

Eval num_timesteps=8000, episode_reward=13.20 +/- 0.40

Episode length: 29.00 +/- 0.00

Step: 8192 Done count: 65

Step: 8704 Done count: 72

Step: 9216 Done count: 76

Step: 9728 Done count: 81

Eval num_timesteps=10000, episode_reward=6.80 +/- 0.40

Episode length: 39.20 +/- 16.10

Step: 10240 Done count: 85

Step: 10752 Done count: 94

Step: 11264 Done count: 102

Step: 11776 Done count: 107

Eval num_timesteps=12000, episode_reward=14.20 +/- 1.83

Episode length: 19.60 +/- 2.24

Step: 12288 Done count: 111

Step: 12800 Done count: 113

Step: 13312 Done count: 118

Step: 13824 Done count: 121

Eval num_timesteps=14000, episode_reward=13.60 +/- 2.87

Episode length: 29.60 +/- 13.66

Step: 14336 Done count: 127

Step: 14848 Done count: 131

Step: 15360 Done count: 139

Step: 15872 Done count: 144

Eval num_timesteps=16000, episode_reward=7.60 +/- 1.50

Episode length: 25.00 +/- 17.40

Step: 16384 Done count: 150

Step: 16896 Done count: 155

Step: 17408 Done count: 158

Step: 17920 Done count: 164

Eval num_timesteps=18000, episode_reward=9.20 +/- 3.19

Episode length: 20.40 +/- 7.23

Step: 18432 Done count: 171

Step: 18944 Done count: 177

Step: 19456 Done count: 182

Step: 19968 Done count: 185

Eval num_timesteps=20000, episode_reward=10.00 +/- 4.52

Episode length: 25.00 +/- 10.81

Step: 20480 Done count: 189

Step: 20992 Done count: 192

Step: 21504 Done count: 196

Step: 22016 Done count: 204

Step: 22528 Done count: 210

Step: 23040 Done count: 214

Step: 23552 Done count: 218

Eval num_timesteps=24000, episode_reward=12.20 +/- 1.72

Episode length: 17.20 +/- 1.72

Step: 24064 Done count: 221

Step: 24576 Done count: 225

Step: 25088 Done count: 231

Step: 25600 Done count: 234

[I 2024-04-25 14:18:29,601] Trial 26 finished with value: 24.2 and parameters: {'n_steps': 268.4160444968466, 'gamma': 0.98, 'learning_rate': 4.2792045832687424e-05, 'ent_coef': 0.00038792901717832427, 'clip_range': 0.2608593226901129, 'n_epochs': 2.3643554800545536}. Best is trial 17 with value: 42.6.
c:\Users\matth\AppData\Local\Programs\Python\Python311\Lib\site-packages\stable_baselines3\ppo\ppo.py:155: UserWarning: You have specified a mini-batch size of 64, but because the `RolloutBuffer` is of size `n_steps * n_envs = 5596`, after every 87 untruncated mini-batches, there will be a truncated mini-batch of size 28
We recommend using a `batch_size` that is a factor of `n_steps * n_envs`.
Info: (n_steps=1399 and n_envs=4)
  warnings.warn(


Output()

Step: 512 Done count: 5

Step: 1024 Done count: 10

Step: 1536 Done count: 14

Eval num_timesteps=2000, episode_reward=28.40 +/- 6.34

Episode length: 34.20 +/- 7.14

New best mean reward!

Step: 2048 Done count: 19

Step: 2560 Done count: 24

Step: 3072 Done count: 29

Step: 3584 Done count: 34

Eval num_timesteps=4000, episode_reward=24.20 +/- 5.53

Episode length: 29.60 +/- 6.18

Step: 4096 Done count: 38

Step: 4608 Done count: 41

Step: 5120 Done count: 45

Step: 5632 Done count: 49

Eval num_timesteps=6000, episode_reward=15.00 +/- 5.33

Episode length: 24.80 +/- 5.74

Step: 6144 Done count: 52

Step: 6656 Done count: 58

Step: 7168 Done count: 62

Step: 7680 Done count: 65

Eval num_timesteps=8000, episode_reward=19.20 +/- 8.33

Episode length: 31.20 +/- 10.15

Step: 8192 Done count: 69

Step: 8704 Done count: 72

Step: 9216 Done count: 83

Step: 9728 Done count: 89

Eval num_timesteps=10000, episode_reward=12.60 +/- 8.75

Episode length: 25.20 +/- 12.14

Step: 10240 Done count: 92

Step: 10752 Done count: 97

Step: 11264 Done count: 105

Step: 11776 Done count: 112

Eval num_timesteps=12000, episode_reward=13.40 +/- 4.13

Episode length: 29.60 +/- 8.78

Step: 12288 Done count: 118

Step: 12800 Done count: 121

Step: 13312 Done count: 128

Step: 13824 Done count: 134

Eval num_timesteps=14000, episode_reward=18.40 +/- 0.80

Episode length: 37.60 +/- 1.36

Step: 14336 Done count: 139

Step: 14848 Done count: 142

Step: 15360 Done count: 146

Step: 15872 Done count: 150

Eval num_timesteps=16000, episode_reward=16.00 +/- 3.90

Episode length: 33.80 +/- 5.98

Step: 16384 Done count: 154

Step: 16896 Done count: 157

Step: 17408 Done count: 161

Step: 17920 Done count: 165

Eval num_timesteps=18000, episode_reward=21.40 +/- 8.71

Episode length: 25.20 +/- 6.88

Step: 18432 Done count: 169

Step: 18944 Done count: 173

Step: 19456 Done count: 179

Step: 19968 Done count: 188

Eval num_timesteps=20000, episode_reward=16.98 +/- 13.07

Episode length: 20.60 +/- 11.48

Step: 20480 Done count: 193

Step: 20992 Done count: 197

Step: 21504 Done count: 203

Eval num_timesteps=22000, episode_reward=13.60 +/- 8.78

Episode length: 17.60 +/- 6.95

Step: 22016 Done count: 209

Step: 22528 Done count: 213

Step: 23040 Done count: 219

Step: 23552 Done count: 226

Eval num_timesteps=24000, episode_reward=30.60 +/- 2.42

Episode length: 31.60 +/- 1.62

New best mean reward!

Step: 24064 Done count: 232

Step: 24576 Done count: 238

Step: 25088 Done count: 242

Step: 25600 Done count: 245

Eval num_timesteps=26000, episode_reward=17.40 +/- 8.45

Episode length: 21.40 +/- 6.50

Step: 26112 Done count: 252

Step: 26624 Done count: 262

Step: 27136 Done count: 265

Step: 27648 Done count: 269

[I 2024-04-25 14:55:05,422] Trial 27 finished with value: 30.6 and parameters: {'n_steps': 1399.2680023410692, 'gamma': 0.9, 'learning_rate': 0.0004844749478513348, 'ent_coef': 0.0037842381393682185, 'clip_range': 0.19211280396774147, 'n_epochs': 1.0024868182827513}. Best is trial 17 with value: 42.6.
c:\Users\matth\AppData\Local\Programs\Python\Python311\Lib\site-packages\stable_baselines3\ppo\ppo.py:155: UserWarning: You have specified a mini-batch size of 64, but because the `RolloutBuffer` is of size `n_steps * n_envs = 2548`, after every 39 untruncated mini-batches, there will be a truncated mini-batch of size 52
We recommend using a `batch_size` that is a factor of `n_steps * n_envs`.
Info: (n_steps=637 and n_envs=4)
  warnings.warn(


Output()

Step: 512 Done count: 2

Step: 1024 Done count: 4

Step: 1536 Done count: 8

Eval num_timesteps=2000, episode_reward=22.74 +/- 10.45

Episode length: 32.00 +/- 17.61

New best mean reward!

Step: 2048 Done count: 11

Step: 2560 Done count: 15

Step: 3072 Done count: 20

Step: 3584 Done count: 23

Eval num_timesteps=4000, episode_reward=17.60 +/- 2.33

Episode length: 37.00 +/- 7.07

Step: 4096 Done count: 29

Step: 4608 Done count: 34

Step: 5120 Done count: 38

Step: 5632 Done count: 41

Eval num_timesteps=6000, episode_reward=10.60 +/- 1.96

Episode length: 27.20 +/- 7.22

Step: 6144 Done count: 45

Step: 6656 Done count: 49

Step: 7168 Done count: 56

Step: 7680 Done count: 62

Eval num_timesteps=8000, episode_reward=13.00 +/- 4.52

Episode length: 45.00 +/- 25.95

Step: 8192 Done count: 66

Step: 8704 Done count: 69

Step: 9216 Done count: 72

Step: 9728 Done count: 74

Eval num_timesteps=10000, episode_reward=10.20 +/- 6.49

Episode length: 33.60 +/- 26.82

Step: 10240 Done count: 80

Step: 10752 Done count: 83

Step: 11264 Done count: 92

Step: 11776 Done count: 98

Eval num_timesteps=12000, episode_reward=10.20 +/- 2.32

Episode length: 21.20 +/- 9.24

Step: 12288 Done count: 104

Step: 12800 Done count: 111

Step: 13312 Done count: 116

Step: 13824 Done count: 122

Eval num_timesteps=14000, episode_reward=7.40 +/- 2.24

Episode length: 24.80 +/- 19.00

Step: 14336 Done count: 126

Step: 14848 Done count: 133

Step: 15360 Done count: 136

Step: 15872 Done count: 141

Eval num_timesteps=16000, episode_reward=15.20 +/- 3.60

Episode length: 35.60 +/- 6.80

Step: 16384 Done count: 145

Step: 16896 Done count: 149

Step: 17408 Done count: 152

Step: 17920 Done count: 158

Eval num_timesteps=18000, episode_reward=28.40 +/- 10.37

Episode length: 30.40 +/- 7.99

New best mean reward!

Step: 18432 Done count: 160

Step: 18944 Done count: 164

Step: 19456 Done count: 169

Step: 19968 Done count: 173

Eval num_timesteps=20000, episode_reward=29.00 +/- 7.95

Episode length: 31.00 +/- 5.59

New best mean reward!

Step: 20480 Done count: 179

Step: 20992 Done count: 183

Step: 21504 Done count: 188

Eval num_timesteps=22000, episode_reward=14.80 +/- 5.91

Episode length: 46.40 +/- 24.01

Step: 22016 Done count: 196

Step: 22528 Done count: 199

Step: 23040 Done count: 203

Step: 23552 Done count: 205

Eval num_timesteps=24000, episode_reward=20.00 +/- 6.32

Episode length: 23.80 +/- 4.71

Step: 24064 Done count: 209

Step: 24576 Done count: 213

Step: 25088 Done count: 218

Step: 25600 Done count: 222

Eval num_timesteps=26000, episode_reward=10.80 +/- 2.40

Episode length: 22.60 +/- 7.99

Step: 26112 Done count: 226

Step: 26624 Done count: 232

Step: 27136 Done count: 238

Step: 27648 Done count: 242

Eval num_timesteps=28000, episode_reward=10.80 +/- 3.49

Episode length: 26.00 +/- 4.56

[I 2024-04-25 15:34:06,611] Trial 28 finished with value: 29.0 and parameters: {'n_steps': 637.6502617322847, 'gamma': 0.9999, 'learning_rate': 0.0001530564457810309, 'ent_coef': 0.02862055317357903, 'clip_range': 0.21957397940070242, 'n_epochs': 1.5580612834160128}. Best is trial 17 with value: 42.6.
c:\Users\matth\AppData\Local\Programs\Python\Python311\Lib\site-packages\stable_baselines3\ppo\ppo.py:155: UserWarning: You have specified a mini-batch size of 64, but because the `RolloutBuffer` is of size `n_steps * n_envs = 1668`, after every 26 untruncated mini-batches, there will be a truncated mini-batch of size 4
We recommend using a `batch_size` that is a factor of `n_steps * n_envs`.
Info: (n_steps=417 and n_envs=4)
  warnings.warn(


Output()

Step: 512 Done count: 5

Step: 1024 Done count: 8

Step: 1536 Done count: 12

Eval num_timesteps=2000, episode_reward=13.80 +/- 4.96

Episode length: 37.60 +/- 23.29

New best mean reward!

Step: 2048 Done count: 16

Step: 2560 Done count: 21

Step: 3072 Done count: 23

Step: 3584 Done count: 28

Eval num_timesteps=4000, episode_reward=9.40 +/- 3.44

Episode length: 26.20 +/- 19.37

Step: 4096 Done count: 33

Step: 4608 Done count: 38

Step: 5120 Done count: 41

Step: 5632 Done count: 45

Eval num_timesteps=6000, episode_reward=13.20 +/- 4.12

Episode length: 36.60 +/- 24.06

Step: 6144 Done count: 50

Step: 6656 Done count: 52

Step: 7168 Done count: 56

Step: 7680 Done count: 62

Eval num_timesteps=8000, episode_reward=10.20 +/- 2.93

Episode length: 23.20 +/- 3.49

Step: 8192 Done count: 68

Step: 8704 Done count: 74

Step: 9216 Done count: 79

Step: 9728 Done count: 84

Eval num_timesteps=10000, episode_reward=10.40 +/- 4.18

Episode length: 22.40 +/- 8.14

Step: 10240 Done count: 90

Step: 10752 Done count: 94

Step: 11264 Done count: 97

Step: 11776 Done count: 102

Eval num_timesteps=12000, episode_reward=12.20 +/- 2.79

Episode length: 22.80 +/- 7.25

Step: 12288 Done count: 105

Step: 12800 Done count: 111

Step: 13312 Done count: 114

Step: 13824 Done count: 117

Eval num_timesteps=14000, episode_reward=11.40 +/- 5.24

Episode length: 21.60 +/- 9.58

Step: 14336 Done count: 120

Step: 14848 Done count: 124

Step: 15360 Done count: 133

Step: 15872 Done count: 138

Eval num_timesteps=16000, episode_reward=8.20 +/- 4.96

Episode length: 18.00 +/- 8.02

Step: 16384 Done count: 146

Step: 16896 Done count: 149

Step: 17408 Done count: 156

Step: 17920 Done count: 159

Eval num_timesteps=18000, episode_reward=17.60 +/- 4.27

Episode length: 39.60 +/- 18.07

New best mean reward!

Step: 18432 Done count: 163

Step: 18944 Done count: 167

Step: 19456 Done count: 170

Step: 19968 Done count: 174

Eval num_timesteps=20000, episode_reward=14.00 +/- 4.05

Episode length: 26.20 +/- 6.62

Step: 20480 Done count: 178

Step: 20992 Done count: 182

Step: 21504 Done count: 187

Eval num_timesteps=22000, episode_reward=12.20 +/- 4.12

Episode length: 23.20 +/- 10.59

Step: 22016 Done count: 192

Step: 22528 Done count: 198

Step: 23040 Done count: 203

Step: 23552 Done count: 208

Eval num_timesteps=24000, episode_reward=13.60 +/- 3.07

Episode length: 25.00 +/- 5.76

Step: 24064 Done count: 212

Step: 24576 Done count: 216

Step: 25088 Done count: 221

Step: 25600 Done count: 223

Eval num_timesteps=26000, episode_reward=10.60 +/- 3.14

Episode length: 19.60 +/- 7.55

Step: 26112 Done count: 229

Step: 26624 Done count: 236

[I 2024-04-25 16:12:57,361] Trial 29 finished with value: 17.6 and parameters: {'n_steps': 417.10382828176597, 'gamma': 0.99, 'learning_rate': 2.111541073383337e-05, 'ent_coef': 2.827020910421531e-05, 'clip_range': 0.2751309421738334, 'n_epochs': 3.644051417068777}. Best is trial 17 with value: 42.6.


Number of finished trials:  30
Best trial:
  Value:  42.6
  Params: 
    n_steps: 1298.0336974151803
    gamma: 0.98
    learning_rate: 0.0001291858786170927
    ent_coef: 0.0016824274086785702
    clip_range: 0.19565974040447376
    n_epochs: 1.2218276926103449


In [30]:
from highway_env.envs import IntersectionEnv

IntersectionEnv.default_config()

{'observation': {'type': 'Kinematics',
  'vehicles_count': 15,
  'features': ['presence', 'x', 'y', 'vx', 'vy', 'cos_h', 'sin_h'],
  'features_range': {'x': [-100, 100],
   'y': [-100, 100],
   'vx': [-20, 20],
   'vy': [-20, 20]},
  'absolute': True,
  'flatten': False,
  'observe_intentions': False},
 'action': {'type': 'DiscreteMetaAction',
  'longitudinal': True,
  'lateral': False,
  'target_speeds': [0, 4.5, 9]},
 'simulation_frequency': 15,
 'policy_frequency': 1,
 'other_vehicles_type': 'highway_env.vehicle.behavior.IDMVehicle',
 'screen_width': 600,
 'screen_height': 600,
 'centering_position': [0.5, 0.6],
 'scaling': 7.15,
 'show_trajectories': False,
 'render_agent': True,
 'offscreen_rendering': False,
 'manual_control': False,
 'real_time_rendering': False,
 'duration': 13,
 'destination': 'o1',
 'controlled_vehicles': 1,
 'initial_vehicle_count': 10,
 'spawn_probability': 0.6,
 'collision_reward': -5,
 'high_speed_reward': 1,
 'arrived_reward': 1,
 'reward_speed_range': [

Best trial:
  Value:  42.6
  Params: 
    n_steps: 1298.0336974151803
    gamma: 0.98
    learning_rate: 0.0001291858786170927
    ent_coef: 0.0016824274086785702
    clip_range: 0.19565974040447376
    n_epochs: 1.2218276926103449